### PR TITLE
Sarkars/unlock in except

### PR DESCRIPTION
**Pull request by sayantan-nervana**

### CODE REVIEW

**Comment by sayantan-nervana** [comment]
-----------------------------------------------------------------------
                 Starting tests
-----------------------------------------------------------------------
TensorFlow version:  1.12.0
Compiler version:  5.4.0 20160609
TensorFlow version installed: 1.12.0 (v1.12.0-0-ga6d8ffae09)
nGraph bridge built with: 1.12.0 (v1.12.0-0-ga6d8ffae09)
nGraph bridge version:  b'0.9.0'
23:55:55: TEST: run-a3c.sh;./tools/calc-tf-a3c-results.py 14 seconds    PASSED
23:56:09: TEST: run-dcgan.sh;./tools/calc-dcgan-results.py 25 seconds   PASSED
23:56:34: TEST: run-densenet_40_12.sh 7 seconds         PASSED
23:56:41: TEST: run-draw.sh;./tools/calc-draw-results.py 27 seconds     PASSED
23:57:08: TEST: run-fasterRCNN.sh;./tools/calc-tf-fasterRCNN-results.py 317 seconds     PASSED
00:02:25: TEST: run-gnmt-de-en.sh 238 seconds   PASSED
00:06:24: TEST: run-gnmt-en-vi.sh 317 seconds   PASSED
00:11:41: TEST: run-inception-resnet-v2.sh;./tools/calc-tf-inceptionresnetv2-results.py 46 seconds      PASSED
00:12:27: TEST: run-inception-v3.sh;./tools/calc-tf-cnn-resnet50_v2_results.py 10 seconds      PASSED
00:12:37: TEST: run-inception-v4.sh;./tools/calc-tf-cnn-resnet50_v2_results.py 17 seconds      PASSED
00:12:54: TEST: run-mask-rcnn.sh 51 seconds     PASSED
00:13:45: TEST: run-mobilenet-v2.sh;./tools/calc-tf-cnn-resnet50_v2_results.py 11 seconds      PASSED
00:13:56: TEST: run-mobilenet-v1.sh;./tools/calc-MobileNet_v1-results.py 11 seconds     PASSED
00:14:07: TEST: run-ncf.sh;./tools/calc-ncf-results.py 14 seconds       PASSED
00:14:21: TEST: run-resnet50-v1.sh;./tools/calc-tf-cnn-resnet50_v2_results.py 11 seconds       PASSED
00:14:32: TEST: run-resnet50-v2.sh;./tools/calc-tf-cnn-resnet50_v2_results.py 9 seconds        PASSED
00:14:41: TEST: run-rfcn.sh;./tools/calc-tf-rfcn-results.py 294 seconds         PASSED
00:19:35: TEST: run-squeezenet_v1_1.sh;./tools/calc-tf-squeezenet-results.py 37 seconds        PASSED
00:20:12: TEST: run-ssd-mobilenet-v1.sh;./tools/calc-tf-ssd-results.py 22 seconds       PASSED
00:20:34: TEST: run-transformer-LT.sh 144 seconds       PASSED
00:22:58: TEST: run-unet.sh;./tools/calc-tf-unet-results.py 9 seconds   PASSED
00:23:07: TEST: run-vgg16.sh 16 seconds         PASSED
00:23:23: TEST: run-wavenet.sh;./tools/calc-wavenet-results.py 35 seconds       PASSED
00:23:58: TEST: run-wide-and-deep.sh;./tools/calc-wide-and-deep-results.py 3 seconds    FAILED
00:24:01: TEST: run-yolo-v2.sh;./tools/calc-yolov2-results.py 33 seconds        PASSED

Total run took 1719 seconds
Failed:  1  Passed:  24
FAILED tests:
TEST: run-wide-and-deep.sh

wide-and-deep seems to be broken with ngraph master as backend

File: test/opexecuter.cpp [hunk 1]
```diff
@@ -409,9 +397,13 @@ void OpExecuter::ExecuteOnNGraph(vector<Tensor>& ngraph_outputs,
   try {
     backend->call(backend->compile(ng_function), ng_op_tensors, ng_ip_tensors);
   } catch (const std::exception& exp) {
+    BackendManager::UnlockBackend(ng_backend_type);
+    NgraphSerialize("unit_test_error_" + test_op_type_ + ".json", ng_function);
     std::cout << "Exception while executing on nGraph " << exp.what()
```
Comment (avijit-nervana) [inline test/opexecuter.cpp]: The execution must stop at this point i.e., once the exception is encountered, this function should re-throw the previous exception. This is an existing bug - but please fix this in this PR.

File: test/opexecuter.cpp [hunk 2]
```diff
@@ -409,9 +397,13 @@ void OpExecuter::ExecuteOnNGraph(vector<Tensor>& ngraph_outputs,
...
               << std::endl;
   } catch (...) {
+    BackendManager::UnlockBackend(ng_backend_type);
+    NgraphSerialize("unit_test_error_" + test_op_type_ + ".json", ng_function);
     std::cout << "Exception while executing on nGraph " << std::endl;
```
Comment (avijit-nervana) [inline test/opexecuter.cpp]: Same comment as above.

File: src/ngraph_utils.cc [hunk 3]
```diff
@@ -201,6 +201,22 @@ Status CheckAxisDimInRange(std::vector<int64> axes, size_t rank) {
...
+  NGRAPH_VLOG(0) << "Serializing graph to: " << file_name << std::endl;
+  std::string js = ngraph::serialize(ng_function, 4);
+  std::ofstream f;
+  f.exceptions(std::ofstream::failbit | std::ofstream::badbit);
+  try {
+    f.open(file_name);
+    f << js;
+    f.close();
+  } catch (std::ofstream::failure& e) {
+    std::cerr << "Exception opening/closing file " << file_name << std::endl;
```
Comment (avijit-nervana) [inline src/ngraph_utils.cc]: This should be sent to the log just like in line 206 above.

File: test/opexecuter.cpp [hunk 4]
```diff
@@ -409,10 +397,13 @@ void OpExecuter::ExecuteOnNGraph(vector<Tensor>& ngraph_outputs,
...
-    std::cout << "Exception while executing on nGraph " << exp.what()
-              << std::endl;
+    BackendManager::UnlockBackend(ng_backend_type);
+    NgraphSerialize("unit_test_error_" + test_op_type_ + ".json", ng_function);
+    FAIL() << "Exception while executing on nGraph " << exp.what();
   } catch (...) {
-    std::cout << "Exception while executing on nGraph " << std::endl;
+    BackendManager::UnlockBackend(ng_backend_type);
+    NgraphSerialize("unit_test_error_" + test_op_type_ + ".json", ng_function);
+    FAIL() << "Exception while executing on nGraph";
```
Comment (sayantan-nervana) [inline test/opexecuter.cpp]: Ah, I see what you mean 👍

**Pull request merged**